packages/web3-eth-accounts/src/index.js

```diff
@@ -287,6 +313,93 @@ function _validateTransactionForSigning(tx) {
     return;
 }
 
+function _handleTxType(tx) {
+    let txType = tx.type !== undefined ? utils.toHex(tx.type) : '0x0';
+    // Taken from https://github.com/ethers-io/ethers.js/blob/2a7ce0e72a1e0c9469e10392b0329e75e341cf18/packages/abstract-signer/src.ts/index.ts#L215
+    const hasEip1559 = (tx.maxFeePerGas !== undefined || tx.maxPriorityFeePerGas !== undefined);
+    if (tx.gasPrice !== undefined && (tx.type === '0x2' || hasEip1559))
+        throw Error("eip-1559 transactions don't support gasPrice");
+    if ((tx.type === '0x1' || tx.type === '0x0') && hasEip1559)
+        throw Error("pre-eip-1559 transaction don't support maxFeePerGas/maxPriorityFeePerGas");
+    
+    if (
+        hasEip1559 ||
+        (
+            (tx.common && tx.common.hardfork && tx.common.hardfork.toLowerCase() === 'london') ||
+            (tx.hardfork && tx.hardfork.toLowerCase() === 'london')
+        )
+    ) {
+        txType = '0x2';
+    } else if (
+        tx.accessList ||
+        (
+            (tx.common && tx.common.hardfork && tx.common.hardfork.toLowerCase() === 'berlin') ||
+            (tx.hardfork && tx.hardfork.toLowerCase() === 'berlin')
+        )
+    ) {
+        txType = '0x1';
+    }
+    
+    return txType
+}
+
+function _handleTxPricing(_this, tx) {
+    return new Promise((resolve, reject) => {
+        try {
+            if (
+                (
+                    tx.type === undefined ||
+                    tx.type === '0x0' ||
+                    tx.type === '0x1'
+                ) && tx.gasPrice !== undefined)
+            {
+                // Legacy transaction, return provided gasPrice
+                resolve({ gasPrice: tx.gasPrice })
+            } else {
+                Promise.all([
+                    _this._ethereumCall.getBlockByNumber(),
+                    _this._ethereumCall.getGasPrice()
+                ]).then(responses => {
+                    const [block, gasPrice] = responses;
+                    if (
+                        (tx.type !== '0x0' && 
```

Comment:
Its already checked https://github.com/ChainSafe/web3.js/pull/4155/files#diff-cf46f2521b1de659430e09e654742a41c3696dbdddc3cd563ca4e2c02d18ffacR352 for legacy Tx types. May be we can remove this double check or is there any use of additional check?

Reply:
So while this is double checking `tx.type`, it's for different reasons:

[Here](https://github.com/ChainSafe/web3.js/blob/572e416e197156032b33a767e1457933afc2a68e/packages/web3-eth-accounts/src/index.js#L357) we are short circuiting and returning `gasPrice` if it's apart of the `tx` object and `tx.type` is `0x0` or `0x1` since these txs will not have the EIP1559 fields and will be using the user provided `gasPrice`

And [here](https://github.com/ChainSafe/web3.js/blob/572e416e197156032b33a767e1457933afc2a68e/packages/web3-eth-accounts/src/index.js#L365) we are verifying we aren't of type `0x0` or `0x1` before we add EIP1559 fields

I admit that this code is a bit convoluted, but it's essentially just a condensed version of Ether's EIP1559 implementation